math/src/field/fields/p448_goldilocks_prime_field.rs

```diff
@@ -0,0 +1,304 @@
+
+        let mut aa = [0u64; 4];
+        let mut bb = [0u64; 4];
+        let mut bbb = [0u64; 4];
+
+        for i in 0..4 {
+            aa[i] = a[i] + a[i + 4];
+            bb[i] = b[i] + b[i + 4];
+            bbb[i] = bb[i] + b[i + 4];
+        }
+
+        let widemul = |a: u64, b: u64| -> u128 { (a as u128) * (b as u128) };
+
+        for i in 0..4 {
+            accum2 = 0;
+
+            for j in 0..=i {
+                accum2 += widemul(a[j], b[i - j]);
+                accum1 += widemul(aa[j], bb[i - j]);
+                accum0 += widemul(a[j + 4], b[i - j + 4]);
+            }
+            for j in (i + 1)..4 {
+                accum2 += widemul(a[j], b[8 - (j - i)]);
+                accum1 += widemul(aa[j], bbb[4 - (j - i)]);
+                accum0 += widemul(a[j + 4], bb[4 - (j - i)]);
+            }
+
+            accum1 -= accum2;
+            accum0 += accum2;
+
+            c[i] = (accum0 as u64) & mask;
+            c[i + 4] = (accum1 as u64) & mask;
+
+            accum0 >>= 56;
+            accum1 >>= 56;
+        }
+
+        accum0 += accum1;
+        accum0 += c[4] as u128;
+        accum1 += c[0] as u128;
+        c[4] = (accum0 as u64) & mask;
+        c[0] = (accum1 as u64) & mask;
+
+        accum0 >>= 56;
+        accum1 >>= 56;
+
+        c[5] += accum0 as u64;
+        c[1] += accum1 as u64;
+
+        U56x8 { limbs: c }
+    }
+
+    fn sub(a: &U56x8, b: &U56x8) -> U56x8 {
+        let co1 = ((1u64 << 56) - 1) * 2;
+        let co2 = co1 - 2;
+
+        let mut limbs = [0u64; 8];
+        for (i, limb) in limbs.iter_mut().enumerate() {
+            *limb =
+                a.limbs[i]
+                    .wrapping_sub(b.limbs[i])
+                    .wrapping_add(if i == 4 { co2 } else { co1 });
+        }
+
+        let mut res = U56x8 { limbs };
+        Self::weak_reduce(&mut res);
+        res
+    }
+
+    fn neg(a: &U56x8) -> U56x8 {
+        let zero = Self::zero();
+        Self::sub(&zero, a)
+    }
+
+    fn inv(a: &U56x8) -> U56x8 {
+        Self::pow(a, P448_GOLDILOCKS_PRIME_FIELD_ORDER - U448::from_u64(2))
+    }
+
+    fn div(a: &U56x8, b: &U56x8) -> U56x8 {
+        let b_inv = Self::inv(b);
+        Self::mul(a, &b_inv)
+    }
+
+    fn eq(a: &U56x8, b: &U56x8) -> bool {
+        a.limbs == b.limbs
+    }
+
+    fn zero() -> U56x8 {
+        U56x8 { limbs: [0u64; 8] }
+    }
+
+    fn one() -> U56x8 {
+        let mut limbs = [0u64; 8];
+        limbs[0] = 1;
+        U56x8 { limbs }
+    }
+
+    fn from_u64(x: u64) -> U56x8 {
+        let mut limbs = [0u64; 8];
+        limbs[0] = x & ((1u64 << 56) - 1);
+        limbs[1] = x >> 56;
+        U56x8 { limbs }
+    }
+
+    fn from_base_type(x: U56x8) -> U56x8 {
+        let mut x = x;
+        Self::weak_reduce(&mut x);
+        x
+    }
+}
+
+impl P448GoldilocksPrimeField {
+    fn weak_reduce(a: &mut U56x8) {
+        let a = &mut a.limbs;
+
+        let mask = (1u64 << 56) - 1;
+        let tmp = a[7] >> 56;
```

Comment:
This may be a dumb question, but if the limbs are U56, wouldn't this always be simply 0?

Reply:
Sometimes, while doing certain computations, we allow the values in limbs to exceed `2^56-1` (here in the case of addition and subtraction) and then we want to reduce them again to the usual base `2^56` form (by carry propagation). This function is used for such cases and therefore `a[7] >> 56` might be non zero.

Reply:
I thought that much, but I wanted to make sure.